### PR TITLE
BACK-7: Allow streams to be assigned to storage node addresses

### PR DESCRIPTION
**Pull request by teogeb**
We can assign storage node addresses for streams. The assignments are saved to the database and can be queried/updated via the REST API.

Added REST endpoints:
- GET /api/v1/storageNodes/$storageNodeAddress/streams
- GET /api/v1/streams/$streamId/storageNodes
- POST /api/v1/streams/$streamId/storageNodes
- DELETE /api/v1/streams/$streamId/storageNodes/$storageNodeAddress

### CODE REVIEW

File: grails-app/controllers/com/unifina/controller/StorageNodeApiController.groovy [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+	@StreamrApi(authenticationLevel = AuthLevel.NONE)
+	def findStreamsByStorageNode(String storageNodeAddress) {
+		List<Stream> streams = storageNodeService.findStreamsByStorageNode(storageNodeAddress);
+		return render(streams*.toSummaryMap() as JSON)
+	}
+
+	@StreamrApi(authenticationLevel = AuthLevel.NONE)
+	def findStorageNodesByStream(String streamId) {
+		List<StreamStorageNode> streamStorageNodes = storageNodeService.findStorageNodesByStream(streamId);
```
Comment (kare) [inline grails-app/controllers/com/unifina/controller/StorageNodeApiController.groovy:42]: Remove `;`'s

File: grails-app/controllers/com/unifina/controller/StorageNodeApiController.groovy [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+class StorageNodeApiController {
+
+	private static final Logger log = Logger.getLogger(StorageNodeApiController)
+
+	StorageNodeService storageNodeService
+	PermissionService permissionService
+
+	@StreamrApi(authenticationLevel = AuthLevel.NONE)
+	def findStreamsByStorageNode(String storageNodeAddress) {
```
Comment (kare) [inline grails-app/controllers/com/unifina/controller/StorageNodeApiController.groovy:35]: `storageNodeAddress` requires validation.

File: grails-app/controllers/com/unifina/controller/StorageNodeApiController.groovy [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+		List<Stream> streams = storageNodeService.findStreamsByStorageNode(storageNodeAddress);
+		return render(streams*.toSummaryMap() as JSON)
+	}
+
+	@StreamrApi(authenticationLevel = AuthLevel.NONE)
+	def findStorageNodesByStream(String streamId) {
```
Comment (kare) [inline grails-app/controllers/com/unifina/controller/StorageNodeApiController.groovy:41]: `streamId` requires validation.

File: grails-app/controllers/com/unifina/controller/StorageNodeApiController.groovy [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+		List<StreamStorageNode> streamStorageNodes = storageNodeService.findStorageNodesByStream(streamId);
+		return render(streamStorageNodes*.toMap() as JSON)
+	}
+
+	@StreamrApi(authenticationLevel = AuthLevel.USER)
+	def addStorageNodeToStream(StorageNodeAddCommand command, String streamId) {
```
Comment (kare) [inline grails-app/controllers/com/unifina/controller/StorageNodeApiController.groovy:47]: `streamId` requires validation.

File: grails-app/controllers/com/unifina/controller/StorageNodeApiController.groovy [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+			} else {
+				throw new NotPermittedException(request.apiUser.username, "Stream", streamId)
+			}
+		} else {
+			throw new ValidationException(command.errors)
+		}
+	}
+
+	@StreamrApi(authenticationLevel = AuthLevel.USER)
+	def removeStorageNodeFromStream(String storageNodeAddress, String streamId) {
```
Comment (kare) [inline grails-app/controllers/com/unifina/controller/StorageNodeApiController.groovy:62]: `storageNodeAddress` and `streamId` requires validation.

**Review by kare — CHANGES_REQUESTED** [review]
Some issues found.

File: grails-app/controllers/com/unifina/controller/StorageNodeApiController.groovy [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+import com.unifina.api.ValidationException
+import com.unifina.api.NotPermittedException
+import com.unifina.utils.EthereumAddressValidator
+import com.unifina.domain.Permission
+import com.unifina.domain.User
+import grails.converters.JSON
+import grails.validation.Validateable
+import org.apache.log4j.Logger
+
+@Validateable
```
[the remark this answers is not in the document]
Comment (teogeb) [reply grails-app/controllers/com/unifina/controller/StorageNodeApiController.groovy:18]: Not needed yet, but let's add later if needed

File: rest-e2e-tests/package.json [hunk 7]
```diff
@@ -16,6 +16,7 @@
     "ethereumjs-util": "6.2.0",
     "form-data": "3.0.0",
     "keythereum": "1.0.4",
+    "lodash": "^4.17.20",
```
[the remark this answers is not in the document]
Comment (teogeb) [reply rest-e2e-tests/package.json:19]: This is generic library that is widely used. But will use exact version.

File: grails-app/controllers/com/unifina/controller/StorageNodeApiController.groovy (hunk 3 above)
Comment (teogeb) [reply grails-app/controllers/com/unifina/controller/StorageNodeApiController.groovy:41]: Let's add validation to streamId once it has specific format (hex-string or ens path)

File: grails-app/controllers/com/unifina/controller/StorageNodeApiController.groovy (hunk 5 above)
Comment (teogeb) [reply grails-app/controllers/com/unifina/controller/StorageNodeApiController.groovy:62]: Let's add validation to streamId once it has specific format (hex-string or ens path)

File: grails-app/controllers/com/unifina/controller/StorageNodeApiController.groovy (hunk 4 above)
Comment (teogeb) [reply grails-app/controllers/com/unifina/controller/StorageNodeApiController.groovy:47]: Let's add validation to streamId once it has specific format (hex-string or ens path)

File: rest-e2e-tests/streamr-api-clients.js [hunk 8]
```diff
@@ -426,7 +453,8 @@ module.exports = (baseUrl, logging) => {
                 products: new Products(options),
                 streams: new Streams(options),
                 subscriptions: new Subscriptions(options),
-                dataunions: new DataUnions(options),
+				dataunions: new DataUnions(options),
+				storagenodes: new StorageNodes(options),
```
[the remark this answers is not in the document]
Comment (teogeb) [reply rest-e2e-tests/streamr-api-clients.js:457]: Fixed

File: rest-e2e-tests/storagenode-api.test.js [hunk 9]
```diff
@@ -0,0 +1,172 @@
+const assert = require("chai").assert;
```
[the remark this answers is not in the document]
Comment (teogeb) [reply rest-e2e-tests/storagenode-api.test.js:1]: Removed from imports.

File: grails-app/domain/com/unifina/domain/StreamStorageNode.groovy [hunk 10]
```diff
@@ -0,0 +1,34 @@
+package com.unifina.domain
+
+import grails.compiler.GrailsCompileStatic
+import groovy.transform.EqualsAndHashCode
+import grails.persistence.Entity
+import com.unifina.utils.EthereumAddressValidator
+
+@Entity
+@EqualsAndHashCode(includes="streamId,storageNodeAddress")
```
[the remark this answers is not in the document]
Comment (teogeb) [reply grails-app/domain/com/unifina/domain/StreamStorageNode.groovy:9]: ToString can be added once it is needed. EqualsAndHashcode is a good for other reasons too, most likely nobody removes that without checking if that is actually needed.

**Review by kare — APPROVED** [review]
LGTM

**Pull request merged**